src/psyclone/psyir/nodes/container.py

```diff
@@ -171,21 +171,85 @@ def node_str(self, colour=True):
     def __str__(self):
         return f"Container[{self.name}]\n"
 
-    def get_routine_psyir(self, name):
-        '''Returns the PSyIR for the routine with the given name, or None
-        if a routine with this name does not exist.
+    def get_routine_psyir(self, name, allow_private=False,
+                          check_wildcard_imports=False):
+        '''
+        Searches the Container for a definition of the named routine.
+
+        NOTE: if the named routine corresponds to a generic interface then this
+        method will return None. You will need to use `resolve_routine` first
+        to find the names of the routines that the interface resolves to.
 
-        :param str name: name of the routine to find.
+        If it is not found and the routine is named in an import statement
+        then the search is continued in the named Container. Failing that,
+        all wildcard imports are checked.
 
-        :returns: the PSyIR Routine instance of the subroutine, or None if
-            there is no routine with that name in this container.
-        :rtype: Union[None, psyclone.psyir.nodes.Routine]
+        :param str name: the name of the Routine for which to search.
+        :param bool allow_private: whether the Routine is permitted to have
+            a visibility of PRIVATE.
+
+        :returns: the PSyIR of the named Routine if found, otherwise None.
+        :rtype: :py:class:`psyclone.psyir.nodes.Routine` | NoneType
 
         '''
-        name = name.lower()
-        for routine in self.walk(Routine):
-            if routine.name.lower() == name:
-                return routine
+        rname = name.lower()
+        # pylint: disable=import-outside-toplevel
+        from psyclone.psyir.nodes.routine import Routine
+        from psyclone.psyir.symbols.symbol import Symbol, SymbolError
+        # pylint: enable=import-outside-toplevel
+        for node in self.children:
+            if isinstance(node, Routine) and node.name.lower() == rname:
+                # Check this routine is public
+                routine_sym = self.symbol_table.lookup(node.name)
+                if (allow_private or
+                        routine_sym.visibility == Symbol.Visibility.PUBLIC):
```

Comment:
```
else:
    return None
```
As presumably Fortran does not let private subroutine names clash with possible wildcard imported ones, so we can stop without searching further? (it may be good to verify this)

Reply:
You're quite right - it seems that Fortran doesn't permit this (I expected that it would). Updated code and comments.